### DISCUSSION TITLE
Dual display extension for Store Commerce

### DISCUSSION THREAD

**Discussion by ViktorVu602** [microsoft/Dynamics365Commerce.Solutions#26] [Q&A] [answered]
Hi experts,

I hope everyone is having a great day. Today, I would like to have a question about dual display extension for Store Commerce.
Currently, I am having a customer created a dual display extension referenced from this document: https://learn.microsoft.com/en-us/dynamics365/commerce/dev-itpro/pos-dual-display-extension
The extension has worked fine on MPOS but as customer is moving to Store Commerce, they tried to create the same extension and it is not having the same result. In particular, while the package is built and installed with no issue, there are some info not being displayed correctly:
![image](https://user-images.githubusercontent.com/122328428/211440093-e4152211-5462-4ddd-be7d-02e1f24a8599.png)
![image](https://user-images.githubusercontent.com/122328428/211440143-e973cdbc-b595-47a6-bc15-9498800e2369.png)
I have also taken some screenshots of customer’s extension:
![image](https://user-images.githubusercontent.com/122328428/211440218-b89d21f1-67b8-4d38-a44f-8b32384d9020.png)
![image](https://user-images.githubusercontent.com/122328428/211440238-435a41ab-dc6d-48f5-99d5-f025c16dacc3.png)
![image](https://user-images.githubusercontent.com/122328428/211440250-ddef25cf-29c8-4c5e-aa85-ebc8d3766f8f.png)
![image](https://user-images.githubusercontent.com/122328428/211440276-e19c3a70-6b9b-439a-bb6c-08653d78a0c1.png)
![image](https://user-images.githubusercontent.com/122328428/211440299-a9b1fd7e-6e26-44b3-bd16-08fcaf927d20.png)
Ask: With this info, I would like to confirm if Store Commerce is currently supporting a dual display extension?
In case it does, may I know if there are any changes customer has to do with their extension for it to be compatible with Store Commerce app?

**Comment by madyke** [comment, marked as answer]
Hi @ViktorVu602 - yes, Store Commerce supports dual display extensions. I discussed with Luke and we couldn't think of any reason why the extension stopped working when transitioning from MPOS to Store Commerce. We'll take a look at the code and see if we can spot anything that may be going wrong.

**Reply by ViktorVu602** [reply]
Hi @madyke  - thank you for your confirmation on Store Commerce dual display support. About customer's solution, in case you want to have a look at it, we can schedule a meeting with customer. Please let me know anytime if it is required,

**Reply by ViktorVu602** [reply]
Hi @madyke - customer have tried altering their manifest,json file and while they could load the extension, however, data is not being shown on the dual display extension.
![image](https://user-images.githubusercontent.com/122328428/214205423-845f8c2e-7b0d-4166-824e-85345382ceb3.png)
May I ask if you can provide some input on this ticket when you have the time?

**Reply by madyke** [reply]
Hello @viktorvu602 - I tested this with our sample dual display extension today. While the doc you mentioned and the published sample are both out of date and using the old RetailSDK, when I updated the imports to the new CommerceSDK and commented out the sections that it doesn't look like the customer is using, everything worked fine for me.

It's difficult to see any differences between the sample and what the customer has done in their extension from just the screenshots of code. Can you provide the customer's `DualDisplayCustomControl.ts` file either here or over email so I can check what exactly is going on there and try testing it on my VM?

**Reply by ViktorVu602** [reply]
Hi @madyke - I will attach customer's DualDisplayCustomControl.ts  and DualDisplayCustomControl.html via email since the file types are not supported here.

**Reply by madyke** [reply]
Hi @ViktorVu602 - Luke and I debugged this today and were able to find a couple of fixes that can be made to get this working:

In `DualDisplayCustomControl.ts`:
- Add this line to the end of `this.cartChangedHandler`, which will update the cart lines in the dual display when the cart is changed:
`this.cartLinesDataList.data = ObjectExtensions.isNullOrUndefined(data.cart) ? [] : data.cart.CartLines;`
- In `onReady`:
  - On line 123, add parenthesis to the cart lines observable object:
  `data: this._cartLinesObservable(),`
  - Move the definition of `dataListRootElem` and the call to the control factory after the call to `ko.applyBindingsToNode`. This was the main reason the extension didn't work, since the customer was trying to get access to that HTML element before it had been bound by knockout.

In `DualDisplayCustomControl.html`:
 - Remove `data-bind="msPosDataList: cartLinesDataList"`, it's no longer needed since the cart lines are added in `onReady`.

**Reply by ViktorVu602** [reply]
Hi @madyke, thank you for providing the solution for this issue. As customer has confirmed the solution worked, we can close this discussion.
Again, I would like to thank you for all the help you provided on this discussion.

Best regards,
Viktor

**Comment by Osamaasif9312** [comment]
Hi @ViktorVu602 & @madyke,

I am also migrating from MPOS to store commerce and after deploying dual display it is showing same screen as above. It will be very helpful if you guys can share dual display extension with me.

Thanks in advance.
